DRAFT: Standard template for On-prem deployment  for  public use (#423)
elastisys/compliantkubernetes: docs/operator-manual/on-prem-standard.md

1. Create a git working folder to store Compliant Kubernetes configurations in a version-controlled manner. Run the following commands from the root of the config repo.

    > **_NOTE:_** the following steps are done from the root of the git repository you created for the cofigurations.

    > **_NOTE:_** You can choose names for your service cluster and workload cluster by changing the values for `SERVICE_CLUSTER` and `WORKLOAD_CLUSTERS` respectively.

    ```console
    export CK8S_CONFIG_PATH=${HOME}/<path>
    export CK8S_ENVIRONMENT_NAME=<my-ck8s-cluster>
    export CK8S_CLOUD_PROVIDER=[exoscale|safespring|citycloud|elastx|aws|baremetal]
    export CK8S_FLAVOR=[dev|prod] # defaults to dev
    export CK8S_PGP_FP=<PGP-fingerprint> # retrieve with gpg --list-secret-keys
    SERVICE_CLUSTER="sc"
    WORKLOAD_CLUSTERS="wc"
    ```
    ```console
    mkdir -p $CK8S_CONFIG_PATH #create config path
    ```

1. Create the domain name.
    You need to create a domain name to access the different services in your environment. You will need to set up the following DNS entries (replace example.com with your domain name).

>     - Point these domains to the workload cluster ingress controller:

Reviewer: :thinking: The Ingress Controller doesn't really exist at this point, but I do understand the need to "do away" with DNS early in the process. Any idea how we can solve this chicken-and-egg problem?

Author: @cristiklein I have already addressed @moule3053's feedback :)

Reviewer: Regarding addressing @moule3053 's feedback, I'm confused. Here is what I see:
![Screenshot from 2022-10-14 12-08-09](https://user-images.githubusercontent.com/1660833/195822305-99208176-778d-4dba-875d-8f5fa1391a17.png)

Regarding supported on-call tools, we currently only support Slack and OpsGenie (see config [here](https://github.com/elastisys/compliantkubernetes-apps/blob/main/config/config/sc-config.yaml#L992)). With code changes, we could support [all of these](https://prometheus.io/docs/alerting/latest/configuration/).

# Standard Template for on-prem Environment

> This document contains the instructions on how to setup a new Compliant Kubernetes on-prem environment.

Reviewer 2: ```suggestion
This document contains instructions on how to set up a new Compliant Kubernetes on-prem environment.
```

## Prerequisites

> https://elastisys.io/compliantkubernetes/operator-manual/capacity-management/

Reviewer: I'm unsure to understand why this line is present.

Also, I mirroring our internal docs, I would start with:

```
1. Check the infra-provider. If in doubt, do a [provider audit](link to audit page).
```

Author: That was by mistake :)

1. Prepare ubuntu-based VMs 

>     If you are using public clouds, you can use [Kubespray](https://github.com/kubernetes-sigs/kubespray) to create the VMs.

Author: Azure  uses [AzureRM(resource manager)](https://github.com/kubernetes-sigs/kubespray/tree/master/contrib/azurerm) instead of Terraform. Maybe that is an exception.

>     If you are using public clouds, you can use [Kubespray](https://github.com/kubernetes-sigs/kubespray) to create the VMs.

Reviewer: Sounds good. Please document it accordingly. Perhaps we can have a bullet list:

If you are using public clouds, you can use the scripts included in Kubespray:

* For Azure, use [AzureRM scripts](link).
* For other clouds, use their respective [Terraform scripts](link).

>     If you are using public clouds, you can use [Kubespray](https://github.com/kubernetes-sigs/kubespray) to create the VMs.

Reviewer 2: I think it would be nice to use Terraform as much as possible. I see that Terraform also has an Azure Provider https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs

>     If you are using public clouds, you can use [Kubespray](https://github.com/kubernetes-sigs/kubespray) to create the VMs.

Reviewer: @moule3053 I agree that it would be great to use Terraform with Azure too. However, Kubespray currently provides no script for that, so I'm worried that: (a) writing one would be quite some undertaking, (b) we wouldn't be able to maintain it.

1. Choose names for your service cluster and workload cluster(s):

    > **_NOTE:_** the following steps are done from the root of the ck8s-ops repository

>     > **_WARNING:_** When using the devbox, your local ck8s-ops repository is used for templates, so make sure it is up to date

Reviewer: This line can be removed.

    ```console
    export CK8S_CONFIG_PATH=${HOME}/<path>
    export CK8S_ENVIRONMENT_NAME=<my-ck8s-cluster>
    export CK8S_CLOUD_PROVIDER=[exoscale|safespring|citycloud|elastx|aws|baremetal]
    export CK8S_FLAVOR=[dev|prod] # defaults to dev
    export CK8S_PGP_FP=<PGP-fingerprint> # retrieve with gpg --list-secret-keys
    SERVICE_CLUSTER="sc"
    WORKLOAD_CLUSTERS="wc"
    ```
    ```console
    mkdir -p $CK8S_CONFIG_PATH #create config path
    ```

1. Create the domain name. 

>     You need to create a domain name to access the different services in your environment. You will need to set up the following DNS entries (replace example.com with your domain).

Reviewer 2: ```suggestion
    You need to create a domain name to access the different services in your environment. You will need to set up the following DNS entries (replace `example.com` with your domain name).
```

    - Point these domains to the workload cluster ingress controller:
        - `*.example.com`
    - Point these domains to the service cluster ingress controller:
        - `*.ops.example.com`
        - `dex.example.com`
        - `grafana.example.com`
        - `harbor.example.com`
        - `opensearch.example.com`

1. Create S3 credentials and add them to `.state/s3cfg.ini`.

> 1. Set up Loadbalancer
> 
> You need to setup two loadbalancers, one for workload cluster and one for service cluster.

Reviewer 2: ```suggestion
1. Set up load balancers

You need to set up two load balancers, one for the workload cluster and one for the service cluster.
```

1. Make sure you have [all necessary tools](https://elastisys.io/compliantkubernetes/operator-manual/getting-started/).

1. Clone the Elastisys Compliant Kubernetes Kubespray repo as follows:

    ```console

>     git clone --recursive https://github.com/elastisys/compliantkubernetes-kubespray

Reviewer: Would we clone in the config repo? In that case, we should use `git submodule` here.

Author: I think we shouldn't clone it in the config repo. I checked how the devops people work and they don't put the ck8s (kubespray and apps) repos in their environments config repo. I think we should follow that and allow the user to clone these repos as they prefer.

    done
    ```

> 1. Clone compliantkubernetes-apps and Install Pre-requisites as follows:

Reviewer 2: ```suggestion
1. Clone compliantkubernetes-apps and install pre-requisites as follows:
```

    ```console

>     https://github.com/elastisys/compliantkubernetes-apps.git

Reviewer: `git clone` I assume?

Same question as above. Do we clone outside the config repo (and then we need to specify where) or do we clone inside the config repo (and then we need to use `git submodule`). What do you think works best?

    cd compliantkubernetes-apps
    ansible-playbook -e 'ansible_python_interpreter=/usr/bin/python3' --ask-become-pass --connection local --inventory 127.0.0.1, get-requirements.yaml
    done
    ```
## Deploying Compliant Kubernetes using Kubespray

### Init Kubespray config in your config path.

    ```console

>     export CK8S_KUBESPRAY_REPOSITORY_PATH=/path/to/compliantkubernetes-kubespray

Reviewer: We should already set the correct path, once we decided where to clone the Kubespray layer.

    for CLUSTER in ${SERVICE_CLUSTER} "{WORKLOAD_CLUSTERS}"; do
        $CK8S_KUBESPRAY_REPOSITORY_PATH/bin/ck8s-kubespray init $CLUSTER $CK8S_CLOUD_PROVIDER $CK8S_PGP_FP
    done
    ```

> For Citycloud and ElastX environments, edit the `ck8s-k8s-cluster-openstack.yaml` file and set the following variables:

Reviewer: ```suggestion
For OpenStack-based cloud providers, edit the `ck8s-k8s-cluster-openstack.yaml` file and set the following variables:
```

- `external_openstack_cloud_controller_extra_args.cluster-name` # change this only if multiple sc or wc clusters are in the same openstack project

```console
for CLUSTER in ${SERVICE_CLUSTER} "${WORKLOAD_CLUSTERS[@]}"; do
  vim $CK8S_CONFIG_PATH/$CLUSTER-config/group_vars/k8s_cluster/ck8s-k8s-cluster-openstack.yaml
done
```

Update `sc-config/group_vars/k8s_cluster/ck8s-k8s-cluster-openstack.yaml`

```yaml
supplementary_addresses_in_ssl_keys: ["<announce ip for sc>"]
```

Update `wc-config/group_vars/k8s_cluster/ck8s-k8s-cluster-openstack.yaml`

```yaml
supplementary_addresses_in_ssl_keys: ["<announce ip for wc>"]
```

### Configure OIDC

> To configure OIDC access for kubernetes api and  other services dex should be configured with your identity provider. Check what dex needs from [your identity provider](https://dexidp.io/docs/connectors/).

Reviewer: Nice! Minor capitalization fixes:

```suggestion
To configure OpenID access for Kubernetes API and other services, Dex should be configured with your identity provider. Check what Dex needs from [your identity provider](https://dexidp.io/docs/connectors/).
```

#### Configure OIDC endpoint

> Set `kube_oidc_url` in `${CK8S_CONFIG_PATH}/sc-config/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml` and `${CK8S_CONFIG_PATH}/wc-config/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml` based on your cluster. For example, if your domain is `example.com` then kube_oidc_url is set as `kube_oidc_url: https://dex.example.com` in the both files.

Reviewer: ```suggestion
Set `kube_oidc_url` in `${CK8S_CONFIG_PATH}/sc-config/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml` and `${CK8S_CONFIG_PATH}/wc-config/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml` based on your cluster. For example, if your domain is `example.com` then kube_oidc_url is set as `kube_oidc_url: https://dex.example.com` in both files.
```

### Copy the VMs information to the inventery files

>  Add the host name,  user and ip address of each VM that you prepared above in `${CK8S_CONFIG_PATH}/sc-config/inventory.ini`for service cluster and `${CK8S_CONFIG_PATH}/sc-config/inventory.ini` for workload cluster. Moreover, you also need to add the host names of the master nodes under `[kube_control_plane]`, etdc nodes under `[etcd]` and worker nodes under `[kube_node]`.  

Reviewer: Isn't the user always going to be `ubuntu`?

Also, aren't we always going to use the same Nodes for `etcd` as for `kube_control_plane`?

```suggestion
 Add the host name, user and IP address of each VM that you prepared above in `${CK8S_CONFIG_PATH}/sc-config/inventory.ini`for service cluster and `${CK8S_CONFIG_PATH}/sc-config/inventory.ini` for workload cluster. Moreover, you also need to add the host names of the master nodes under `[kube_control_plane]`, etdc nodes under `[etcd]` and worker nodes under `[kube_node]`.  
```

>   > **_NOTE:_** Make sure that the user has ssh access to the VMs.

Reviewer: ```suggestion
  > **_NOTE:_** Make sure that the user has SSH access to the VMs.
```

> ### Run kubespray to deploy the Kubernetes clusters

Reviewer 2: ```suggestion
### Run Kubespray to deploy the Kubernetes clusters
```

> **_NOTE:_** the client openstack credentials will be used for this step

```console
for CLUSTER in ${SERVICE_CLUSTER} ${WORKLOAD_CLUSTERS}; do
    $CK8S_KUBESPRAY_REPOSITORY_PATH/bin/ck8s-kubespray apply $CLUSTER --flush-cache 
done
```

> **_NOTE:_** the kubeconfig for wc `.state/kube_config_wc.yaml` will not be usable until you have installed dex in the service cluster (by deploying apps).

### Set up Rook

_Only for cloud providers that doesn't natively support storage kubernetes._

Run the following command to set up Rook.

```console
 for CLUSTER in  sc wc; do
     sops --decrypt ${CK8S_CONFIG_PATH}/.state/kube_config_$CLUSTER.yaml > $CLUSTER.yaml
     export KUBECONFIG=$CLUSTER.yaml
     $CK8S_KUBESPRAY_REPOSITORY_PATH/compliantkubernetes-kubespray/rook/deploy-rook.sh
 done
```

> Please restart the operator pod, `rook-ceph-operator*`, if some pods stalls in initialization state as shown below:

Reviewer 2: ```suggestion
Please restart the operator pod, `rook-ceph-operator*`, if some pods stall in the initialization state as shown below:
```

```console
rook-ceph     rook-ceph-crashcollector-minion-0-b75b9fc64-tv2vg    0/1     Init:0/2   0          24m
rook-ceph     rook-ceph-crashcollector-minion-1-5cfb88b66f-mggrh   0/1     Init:0/2   0          36m
rook-ceph     rook-ceph-crashcollector-minion-2-5c74ffffb6-jwk55   0/1     Init:0/2   0          14m
```

!!!important
    Pods in pending state usually indicate resource shortage. In such cases you need to use bigger instances.
<!--deploy-rook-stop-->

<!--test-rook-start-->
### Test Rook

To test Rook, proceed as follows:

```bash
for CLUSTER in ${SERVICE_CLUSTER} "${WORKLOAD_CLUSTERS[@]}"; do
    sops exec-file ${CK8S_CONFIG_PATH}/.state/kube_config_$CLUSTER.yaml 'kubectl --kubeconfig {} apply -f https://raw.githubusercontent.com/rook/rook/release-1.5/cluster/examples/kubernetes/ceph/csi/rbd/pvc.yaml';
done

for CLUSTER in ${SERVICE_CLUSTER} "${WORKLOAD_CLUSTERS[@]}"; do
    sops exec-file ${CK8S_CONFIG_PATH}/.state/kube_config_$CLUSTER.yaml 'kubectl --kubeconfig {} get pvc';
done
```

> You should see PVCs in Bound state. If you want to clean the previously created PVCs:

Reviewer 2: ```suggestion
You should see PVCs in `Bound` state. If you want to clean the previously created PVCs:
```

```bash
for CLUSTER in ${SERVICE_CLUSTER} "${WORKLOAD_CLUSTERS[@]}"; do
    sops exec-file ${CK8S_CONFIG_PATH}/.state/kube_config_$CLUSTER.yaml 'kubectl --kubeconfig {} delete pvc rbd-pvc';
done
```
## Deploying Compliant Kubernetes Apps

### Initialize the apps configuration

```bash

> export  CK8S_APPS_REPOSITORY_PATH=</path/to/compliantkubernetes-kubespray-apps>

Reviewer: Once we decided above where to clone the apps layer, we can already set the right path here.

$CK8S_APPS_REPOSITORY_PATH/bin/ck8s init
```

This will initialise the configuration in the `${CK8S_CONFIG_PATH}` directory. Generating configuration files `common-config.yaml`, `sc-config.yaml` and `wc-config.yaml`, as well as secrets with randomly generated passwords in `secrets.yaml`. This will also generate read-only default configuration under the directory `defaults/` which can be used as a guide for available and suggested options.

### Configure the apps and secrets

The configuration files contain some predefined values. You may want to check and edit based on your current environment requirements. The configuration files that require editing are `${CK8S_CONFIG_PATH}/common-config.yaml`, `${CK8S_CONFIG_PATH}/sc-config.yaml`, `${CK8S_CONFIG_PATH}/wc-config.yaml` and `${CK8S_CONFIG_PATH}/secrets.yaml` and set the appropriate values for some of the configuration fields.
Note that, the latter is encrypted.

```console
vim ${CK8S_CONFIG_PATH}/sc-config.yaml

vim ${CK8S_CONFIG_PATH}/wc-config.yaml

vim ${CK8S_CONFIG_PATH}/common-config.yaml
```
Edit the secrets.yaml file and add the credentials for:
- s3 - used for backup storage

> - dex - connectors -- check [your indentiy provider](https://dexidp.io/docs/connectors/).

Reviewer 2: Make this multi-line?

Reviewer: Please also add the On-Call Management Tool here.

```console
sops ${CK8S_CONFIG_PATH}/secrets.yaml
```

> **_TIP:_** The default configuration for the service cluster and workload cluster are available in the directory `${CK8S_CONFIG_PATH}/defaults/` and can be used as a reference for available options.

> **_WARNING:_** Do not modify the read-only default configurations files found in the directory `${CK8S_CONFIG_PATH}/defaults/`. Instead configure the cluster by modifying the regular files `${CK8S_CONFIG_PATH}/sc-config.yaml` and `${CK8S_CONFIG_PATH}/wc-config.yaml` as they will override the default options.

### Create S3 buckets

> You can use the following command to create required S3 buckets. The command uses s3cmd in the background and gets configuration and credentials for your S3 provider from the `~/.s3cfg` file.

Reviewer 2: ```suggestion
You can use the following command to create the required S3 buckets. The command uses `s3cmd` in the background and gets configuration and credentials for your S3 provider from the `~/.s3cfg` file.
```

 ```console
$CK8S_APPS_REPOSITORY_PATH/bin/ck8s s3cmd create
```

### Install Compliant Kubernetes apps

This will set up apps, first in the service cluster and then in the workload cluster:

```console
$CK8S_APPS_REPOSITORY_PATH/compliantkubernetes-apps/bin/ck8s apply sc
$CK8S_APPS_REPOSITORY_PATH/compliantkubernetes-apps/bin/ck8s apply wc

> done

Reviewer: I guess we need to remove `done`.

```

### Settling

> **_INFO:_** Leave sufficient time for the system to settle, e.g., request TLS certificates from LetsEncrypt, perhaps as much as 20 minutes.

Check if all helm charts succeeded.

```console
$CK8S_APPS_REPOSITORY_PATH/compliantkubernetes-apps/bin/ck8s ops helm wc list -A --all
```

You can check if the system settled as follows.

```console
for CLUSTER in ${SERVICE_CLUSTER} "${WORKLOAD_CLUSTERS[@]}"; do
 $CK8S_APPS_REPOSITORY_PATH/compliantkubernetes-apps/bin/ck8s ops kubectl $CLUSTER get --all-namespaces pods
done
```

> Check the output of the command above. All Pods needs to be Running or Completed.

Reviewer 2: ```suggestion
Check the output of the command above. All Pods need to be `Running` or `Completed` status.
```

```console
for CLUSTER in ${SERVICE_CLUSTER} "${WORKLOAD_CLUSTERS[@]}"; do
  $CK8S_APPS_REPOSITORY_PATH/compliantkubernetes-apps/bin/ck8s ops kubectl $CLUSTER get --all-namespaces issuers,clusterissuers,certificates
done
```

Check the output of the command above.

> All resources need to have the Ready column True.

Reviewer 2: ```suggestion
All resources need to have the `Ready` column `True`.
```
